### PR TITLE
MC-169 : Add a command to purge database of mapping

### PR DESCRIPTION
**Pull request by willy-ahva**
Bug fix: n
Feature addition: y
Backwards compatibility break: n
Phpspec specification passes: y
Checkstyle issues: n
ChangeLog updated: n
Fixes the following issue: MC-169

### CODE REVIEW

File: Helper/RemoveMappingFromDBHelper.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+     * @param string $jobInstanceCode
+     *
+     * @return int See class constants (0 = done, 1 = job instance not found, 2 = done but no row affected)
+     */
+    public function removeMapping($jobInstanceCode)
+    {
+        $jobInstance = $this->jobInstanceRepository->findOneByCode($jobInstanceCode);
+
+        if (null === $jobInstance) {
+            return static::JOB_INSTANCE_NOT_FOUND;
```
Comment (fitn) [inline Helper/RemoveMappingFromDBHelper.php]: Exception

File: Command/RemoveMappingFromDBCommand.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        $jobInstanceCode = $input->getArgument('job_instance_code');
+
+        $output->writeln(sprintf('<info>Executing command for %s job instance.<info>', $jobInstanceCode));
+
+        try {
+            $removeMappingService->removeMapping($jobInstanceCode);
+            $output->writeln(
+                sprintf('<info>Mapping related to %s job instance has been removed.<info>', $jobInstanceCode)
+            );
+        } catch (EntityNotFoundException $e) {
```
Comment (fitn) [inline Command/RemoveMappingFromDBCommand.php]: \Exception

File: Command/RemoveMappingFromDBCommand.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            $output->writeln(
+                sprintf('<error>Job instance %s hasn\'t been found, verify your code.<error>', $jobInstanceCode)
```
Comment (fitn) [inline Command/RemoveMappingFromDBCommand.php]: <error>Error appears for JobCode "myCode": "$e->getMessage()"</error>

File: Helper/RemoveMappingFromDBHelper.php [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+use Pim\Bundle\ImportExportBundle\Entity\Repository\JobInstanceRepository;
+
+/**
+ * Remove mapping from database helper
+ *
+ * @author    Willy Mesnage <willy.mesnage@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class RemoveMappingFromDBHelper
```
Comment (fitn) [inline Helper/RemoveMappingFromDBHelper.php]: MappingRemover

File: Command/RemoveMappingFromDBCommand.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Remove mapping database for a given job instance code
+ *
+ * @author    Willy Mesnage <willy.mesnage@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class RemoveMappingFromDBCommand extends ContainerAwareCommand
```
Comment (fitn) [inline Command/RemoveMappingFromDBCommand.php]: RemoveMappingCommand

File: Command/RemoveMappingFromDBCommand.php [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function configure()
+    {
+        $this
+            ->setName('magento-connector:remove:mapping')
```
Comment (fitn) [inline Command/RemoveMappingFromDBCommand.php]: magento-connector:mapping:remove

File: Command/RemoveMappingFromDBCommand.php [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+        return 0;
+    }
+
+    /**
+     * Get the remove mapping from db helper service
+     *
+     * @return RemoveMappingFromDBHelper
+     */
+    protected function getRemoveMappingFromDBService()
```
Comment (fitn) [inline Command/RemoveMappingFromDBCommand.php]: getMappingRemover

File: Command/RemoveMappingFromDBCommand.php [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                'From which job instance want you remove ?'
+            );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $removeMappingService = $this->getRemoveMappingFromDBService();
```
Comment (fitn) [inline Command/RemoveMappingFromDBCommand.php]: $mappingRemover

File: Helper/RemoveMappingFromDBHelper.php [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    /**
+     * Remove mapping from database
+     *
+     * @param string $jobInstanceCode
+     *
+     * @throws EntityNotFoundException
+     */
+    public function removeMapping($jobInstanceCode)
```
Comment (fitn) [inline Helper/RemoveMappingFromDBHelper.php]: purge

File: Helper/RemoveMappingFromDBHelper.php [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+        $this->executeRemoving($jobInstance->getRawConfiguration());
+    }
+
+    /**
+     * Execute removing
+     *
+     * @param array $rowConfiguration
+     */
+    protected function executeRemoving(array $rowConfiguration)
```
Comment (fitn) [inline Helper/RemoveMappingFromDBHelper.php]: Purge mapping

File: Helper/RemoveMappingFromDBHelper.php [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+     */
+    public function removeMapping($jobInstanceCode)
+    {
+        $jobInstance = $this->jobInstanceRepository->findOneByCode($jobInstanceCode);
+
+        if (null === $jobInstance) {
+            throw new EntityNotFoundException();
+        }
+
+        $this->executeRemoving($jobInstance->getRawConfiguration());
```
Comment (fitn) [inline Helper/RemoveMappingFromDBHelper.php]: put the foreach loop here

File: Resources/config/helpers.yml [hunk 12]
```diff
@@ -0,0 +1,11 @@
+parameters:
+    pim_magento_connector.helper.remove_db_mapping.class: Pim\Bundle\MagentoConnectorBundle\Helper\RemoveMappingFromDBHelper
```
Comment (fitn) [inline Resources/config/helpers.yml]: pim_magento_connector.purger.mapping.class

File: Resources/config/helpers.yml [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+
+services:
+    pim_magento_connector.helper.remove_db_mapping:
```
Comment (fitn) [inline Resources/config/helpers.yml]: pim_magento_connector.purger.mapping

File: Command/PurgeMappingCommand.php [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        $mappingPurger   = $this->getMappingPurgerService();
+        $jobInstanceCode = $input->getArgument('job_instance_code');
+
+        $output->writeln(sprintf('<info>Executing command for %s job instance.<info>', $jobInstanceCode));
+
+        try {
+            $mappingPurger->purge($jobInstanceCode);
+            $output->writeln(
+                sprintf('<info>Mapping related to %s job instance has been purge.<info>', $jobInstanceCode)
```
Comment (fitn) [inline Command/PurgeMappingCommand.php]: Add " between %s
Same line 52

File: Purger/MappingPurger.php [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+     * @param string
+     */
+    public function setMappingClasses()
+    {
+        $classes = func_get_args();
+
+        foreach ($classes as $class) {
+            $this->mappingClasses[] = $class;
+        }
+    }
```
Comment (fitn) [inline Purger/MappingPurger.php]: setMappingClasses(array $mappingClasses)
{
    $this->mappingClasses = $mappingClasses;
}

File: Purger/MappingPurger.php [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+    /**
+     * Execute removing
+     *
+     * @param array  $rowConfiguration
+     * @param string $class
+     */
+    protected function purgeMapping(array $rowConfiguration, $class)
+    {
+        $soapUrl = $rowConfiguration['magentoUrl'] . $rowConfiguration['wsdlUrl'];
```
Comment (fitn) [inline Purger/MappingPurger.php]: Should be in parameter to avoid to recalculate each time ;)

File: Purger/MappingPurger.php [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+
+        $qb = $this->entityManager->createQueryBuilder();
+        $qb->delete($class, 'c')
+            ->where($qb->expr()->eq('c.magentoUrl', ':magentoUrl'))
+            ->setParameter(':magentoUrl', $soapUrl)
+            ->getQuery()
+            ->execute();
+        $this->entityManager->clear();
```
Comment (fitn) [inline Purger/MappingPurger.php]: Is it really necessary? I think you can remove this line

File: Purger/MappingPurger.php (hunk 15 above)
Comment (willy-ahva) [inline Purger/MappingPurger.php]: Nope, argument is not an array but, several string. I need func_get_args

Comment (fitn) [inline Purger/MappingPurger.php]: So set an array!

**Pull request merged**